docs/abi.rst

```diff
@@ -0,0 +1,652 @@
+        Assert(get_volume_of_rectangular_prism(length, width, height) > Int(0))
+    )
+
+Subroutines that Return ABI Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning::
+  :any:`ABIReturnSubroutine` is still taking shape and is subject to backwards incompatible changes.
+
+  * For ABI Application entry point definition, feel encouraged to use :any:`ABIReturnSubroutine`.  Expect a best-effort attempt to minimize backwards incompatible changes along with a migration path.
+  * For general purpose usage, use at your own risk.  Based on feedback, the API and usage patterns will change more freely and with less effort to provide migration paths.
+
+In addition to accepting ABI types as arguments, it's also possible for a subroutine to return an ABI type value.
+
+As mentioned in the Computed Value section (TODO: link), operations which return ABI values instead of traditional :code:`Expr` objects need extra care. In order to solve this problem for subroutines, a new decorator, :any:`@ABIReturnSubroutine <ABIReturnSubroutine>` has been introduced.
+
+The :code:`@ABIReturnSubroutine` decorator should be used with subroutines that return an ABI value. Subroutines defined with this decorator will have two places to output information: the function return value, and a new keyword-only argument called :code:`output`. The function return value must remain an :code:`Expr`, while the :code:`output` keyword argument will contain the ABI value the subroutine wishes to return. An example is below:
+
+.. code-block:: python
+
+    @ABIReturnSubroutine
+    def get_account_status(
+        account: abi.Account, *, output: abi.Tuple2[abi.Uint64, abi.Bool]
+    ) -> Expr:
+        balance = abi.Uint64()
+        is_admin = abi.Bool()
+        return Seq(
+            balance.set(App.localGet(account.address(), Bytes("balance"))),
+            is_admin.set(App.localGet(account.address(), Bytes("is_admin"))),
+            output.set(balance, is_admin),
+        )
+
+    account = abi.make(abi.Address)
+    status = abi.make(abi.Tuple2[abi.Uint64, abi.Bool])
+    program = Seq(
+        account.set(Txn.sender()),
+        # NOTE! The return value of get_account_status(account) is actually a ComputedValue[abi.Tuple2[abi.Uint64, abi.Bool]]
+        get_account_status(account).store_into(status),
+    )
+
+Notice that even though the original :code:`get_account_status` function returns an :code:`Expr` object, the :code:`@ABIReturnSubroutine` decorator automatically transforms the function's return value and the :code:`output` variable into a :code:`ComputedValue`. As a result, callers of this subroutine must work with a :code:`ComputedValue`.
+
+The only exception to this transformation is if the subroutine has no return value, in which case a :code:`ComputedValue` is unnecessary, so the subroutine will still return an :code:`Expr` to the caller. In this case the :code:`@ABIReturnSubroutine` decorator acts identically the original :code:`@Subroutine` decorator.
```

Comment:
@jasonpaulos Minor and optional _suggestions_ based on somewhat subjective preferences for shorter sentences and present tense usage.

```suggestion
The only exception to this transformation is if the subroutine has no return value. Without a return value, :code:`ComputedValue` is unnecessary and the subroutine returns an :code:`Expr` to the caller. In this case, the :code:`@ABIReturnSubroutine` decorator acts identically the original :code:`@Subroutine` decorator.
```